### PR TITLE
feat(dc/interface): add new resource support

### PR DESCRIPTION
**Pull request by Lance52259**
**What this PR does / why we need it**:
Support new resource: huaweicloud_dc_virtual_interface

**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
fixes #xxx

**Special notes for your reviewer**:

**Release note**:

```release-note
1. add new resource support.
2. add related document and acc test.
```

## PR Checklist

* [x] Tests added/passed.
* [x] Documentation updated.
* [x] Schema updated.

## Acceptance Steps Performed

```
make testacc TEST='./huaweicloud/services/acceptance/dc' TESTARGS='-run=TestAccVirtualInterface_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/dc -v -run=TestAccVirtualInterface_basic -timeout 360m -parallel 4
=== RUN   TestAccVirtualInterface_basic
=== PAUSE TestAccVirtualInterface_basic
=== CONT  TestAccVirtualInterface_basic
--- PASS: TestAccVirtualInterface_basic (49.48s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/dc        49.577s
```

### CODE REVIEW

File: huaweicloud/services/dc/resource_huaweicloud_dc_virtual_interface.go [hunk 1]
```diff
@@ -0,0 +1,458 @@
...
+	if err != nil {
+		return diag.Errorf("error creating DC v3 client: %s", err)
+	}
+
+	interfaceId := d.Id()
+	resp, err := interfaces.Get(client, interfaceId)
+	if err != nil {
+		return common.CheckDeletedDiag(d, err, "virtual interface")
+	}
+	log.Printf("[Lance] The response of virtual interface is: %#v", resp)
```
Comment (hmbig2) [inline huaweicloud/services/dc/resource_huaweicloud_dc_virtual_interface.go]:
```suggestion
	log.Printf("[Debug] The response of virtual interface is: %#v", resp)
```

File: huaweicloud/services/dc/resource_huaweicloud_dc_virtual_interface.go [hunk 2]
```diff
@@ -0,0 +1,458 @@
...
+
+func updateVirtualInterface(client *golangsdk.ServiceClient, interfaceId string, opts interfaces.UpdateOpts) error {
+	_, err := interfaces.Update(client, interfaceId, opts)
+	if err != nil {
+		return fmt.Errorf("error updating virtual interface (%s): %s", interfaceId, err)
+	}
+	return nil
+}
+
+func updateVirtualInterfaceBasicInfo(client *golangsdk.ServiceClient, d *schema.ResourceData) error {
```
Comment (hmbig2) [inline huaweicloud/services/dc/resource_huaweicloud_dc_virtual_interface.go]: suggest merge updateVirtualInterface and updateVirtualInterfaceBasicInfo to one function.
or change one function which build the params

File: huaweicloud/services/dc/resource_huaweicloud_dc_virtual_interface.go [hunk 3]
```diff
@@ -0,0 +1,458 @@
...
+	if d.HasChanges("name", "description", "bandwidth", "remote_ep_group") {
+		if err = updateVirtualInterfaceBasicInfo(client, d); err != nil {
+			return diag.FromErr(err)
+		}
+	}
+	if d.HasChanges("enable_bfd", "enable_nqa") {
+		if err = closeVirtualInterfaceNetworkDetection(client, d); err != nil {
+			return diag.FromErr(err)
+		}
+		if err = openVirtualInterfaceNetworkDetection(client, d); err != nil {
```
Comment (hmbig2) [inline huaweicloud/services/dc/resource_huaweicloud_dc_virtual_interface.go:419]: it‘s very confusing here. 
maybe you can return two opts 
then invoke interfaces.Update(client, interfaceId, opts)

File: huaweicloud/services/dc/resource_huaweicloud_dc_virtual_interface.go (hunk 1 above)
Comment (Lance52259) [reply huaweicloud/services/dc/resource_huaweicloud_dc_virtual_interface.go]: fixed.

File: huaweicloud/services/dc/resource_huaweicloud_dc_virtual_interface.go (hunk 2 above)
Comment (Lance52259) [reply huaweicloud/services/dc/resource_huaweicloud_dc_virtual_interface.go]: If so, the funtion will hard to read.
I rename the function name: `updateVirtualInterfaceBasicParams`

File: huaweicloud/services/dc/resource_huaweicloud_dc_virtual_interface.go (hunk 3 above)
Comment (Lance52259) [reply huaweicloud/services/dc/resource_huaweicloud_dc_virtual_interface.go:419]: Added some comments to describe these functions call.

File: huaweicloud/services/dc/resource_huaweicloud_dc_virtual_interface.go (hunk 2 above)
Comment (Lance52259) [reply huaweicloud/services/dc/resource_huaweicloud_dc_virtual_interface.go]: done.

File: docs/resources/dc_virtual_interface.md [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+
+* `route_mode` - (Required, String, ForceNew) Specifies the route mode of the virtual interface.  
+  The valid values are **static** and **bgp**.  
+  Changing this will create a new resource.
+
+* `vlan` - (Required, Int, ForceNew) Specifies the VLAN for constomer side.  
+  The valid value is range from `0` to `3,999`.
+  Changing this will create a new resource.
+
+* `bandwidth` - (Required, String) Specifies the type of the virtual interface.
```
Comment (Jason-Zhang9309) [inline docs/resources/dc_virtual_interface.md]: the type is int in the code

File: docs/resources/dc_virtual_interface.md [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+
+  -> Exactly one of `local_gateway_v4_ip` and `local_gateway_v6_ip` must be set.
+
+* `remote_gateway_v4_ip` - (Optional, String, ForceNew) Specifies the IPv4 address of the virtual interface in client
+  side.  
+  Required if `local_gateway_v4_ip` is set.
+  Changing this will create a new resource.
+
+* `address_family` - (Optional, String, ForceNew) Specifies the service type of the virtual interface.  
+  The valid values are **VPC**, **VGW**, **GDWW** and **LGW**.
```
Comment (Jason-Zhang9309) [inline docs/resources/dc_virtual_interface.md]: the values not correct

**Comment by Lance52259** [comment]
```
make testacc TEST='./huaweicloud/services/acceptance/dc' TESTARGS='-run=TestAccVirtualInterface_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/dc -v -run=TestAccVirtualInterface_basic -timeout 360m -parallel 4
=== RUN   TestAccVirtualInterface_basic
=== PAUSE TestAccVirtualInterface_basic
=== CONT  TestAccVirtualInterface_basic
--- PASS: TestAccVirtualInterface_basic (47.79s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/dc        47.863s
```

File: docs/resources/dc_virtual_interface.md (hunk 5 above)
Comment (Lance52259) [reply docs/resources/dc_virtual_interface.md]: fixed.

File: docs/resources/dc_virtual_interface.md (hunk 4 above)
Comment (Lance52259) [reply docs/resources/dc_virtual_interface.md]: fixed.

File: docs/resources/dc_virtual_interface.md [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+
+* `route_mode` - (Required, String, ForceNew) Specifies the route mode of the virtual interface.  
+  The valid values are **static** and **bgp**.  
+  Changing this will create a new resource.
+
+* `vlan` - (Required, Int, ForceNew) Specifies the VLAN for constomer side.  
+  The valid value is range from `0` to `3,999`.
+  Changing this will create a new resource.
+
+* `bandwidth` - (Required, Int) Specifies the type of the virtual interface.
```
Comment (ShiChangkuo) [inline docs/resources/dc_virtual_interface.md]: it seems that the description of `bandwidth` is not correct, please fix it.

File: docs/resources/dc_virtual_interface.md [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+  A CIDR that contains CIDRs of local subnet (corresponding to the parameter `local_gateway_v4_ip` or
+  `local_gateway_v6_ip`) and remote subnet (corresponding to the parameter `remote_gateway_v4_ip` or
+  `remote_gateway_v6_ip`) must exist in the list.
+
+* `description` - (Optional, String) Specifies the description of the virtual interface.  
+  The description contain a maximum of `128` characters and the angle brackets (< and >) are not allowed.  
+  Chinese characters must be in **UTF-8** or **Unicode** format.
+
+* `service_type` - (Optional, String, ForceNew) Specifies the service type of the virtual interface.  
+  The valid values are **vpc**, **VGW**, **GDWW** and **LGW**.
```
Comment (ShiChangkuo) [inline docs/resources/dc_virtual_interface.md]: it's better to add the default value.

File: huaweicloud/services/dc/resource_huaweicloud_dc_virtual_interface.go [hunk 8]
```diff
@@ -0,0 +1,442 @@
...
+				Type:         schema.TypeInt,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.IntBetween(0, 3999),
+				Description:  "The VLAN for constom side.",
+			},
+			"bandwidth": {
+				Type:         schema.TypeInt,
+				Required:     true,
+				ValidateFunc: validation.IntBetween(2, 2147483647),
```
Comment (ShiChangkuo) [inline huaweicloud/services/dc/resource_huaweicloud_dc_virtual_interface.go]: please check the max value of bandwidth

File: huaweicloud/services/dc/resource_huaweicloud_dc_virtual_interface.go [hunk 9]
```diff
@@ -0,0 +1,442 @@
...
+		opts        = interfaces.UpdateOpts{}
+	)
+
+	// At the same time, only one of BFD and NQA is enabled.
+	if d.HasChange("enable_bfd") && !d.Get("enable_bfd").(bool) {
+		opts.EnableBfd = utils.Bool(false)
+	} else if d.HasChange("enable_nqa") && !d.Get("enable_nqa").(bool) {
+		opts.EnableNqa = utils.Bool(false)
+	}
+	if reflect.DeepEqual(opts, interfaces.UpdateOpts{}) {
```
Comment (ShiChangkuo) [inline huaweicloud/services/dc/resource_huaweicloud_dc_virtual_interface.go:348]:
```suggestion
	} else {
```

File: huaweicloud/services/dc/resource_huaweicloud_dc_virtual_interface.go [hunk 10]
```diff
@@ -0,0 +1,442 @@
...
+		_, err := interfaces.Update(client, interfaceId, opts)
+		if err != nil {
+			return diag.Errorf("error closing network detection of the virtual interface (%s): %s", interfaceId, err)
+		}
+	}
+	if d.HasChanges("enable_bfd", "enable_nqa") {
+		// BFD and NQA cannot be enabled at the same time.
+		// When BFD (NQA) is enabled and NQA (BFD) is disabled, we need to disable BFD (NQA) first, and then enable NQA (BFD).
+		// If the disable and enable requests are sent at the same time, an error will be reported.
+		if err = closeVirtualInterfaceNetworkDetection(client, d); err != nil {
```
Comment (ShiChangkuo) [inline huaweicloud/services/dc/resource_huaweicloud_dc_virtual_interface.go:416]: you can merge the two functions with
```
opts.EnableNqa = utils.Bool(d.Get("enable_bfd").(bool) == true)
```

File: docs/resources/dc_virtual_interface.md (hunk 7 above)
Comment (Lance52259) [reply docs/resources/dc_virtual_interface.md]: done. The default value is **VGW**.

File: huaweicloud/services/dc/resource_huaweicloud_dc_virtual_interface.go (hunk 8 above)
Comment (Lance52259) [reply huaweicloud/services/dc/resource_huaweicloud_dc_virtual_interface.go]: The validation has been removed.

File: docs/resources/dc_virtual_interface.md (hunk 6 above)
Comment (Lance52259) [reply docs/resources/dc_virtual_interface.md]: fixed.

File: huaweicloud/services/dc/resource_huaweicloud_dc_virtual_interface.go (hunk 9 above)
Comment (Lance52259) [reply huaweicloud/services/dc/resource_huaweicloud_dc_virtual_interface.go:348]: The create request and update request of the NQA and BFD cannot be sent at the same time.

File: huaweicloud/services/dc/resource_huaweicloud_dc_virtual_interface.go (hunk 10 above)
Comment (Lance52259) [reply huaweicloud/services/dc/resource_huaweicloud_dc_virtual_interface.go:416]: The create request and update request of the NQA and BFD cannot be sent at the same time.

**Pull request merged**